projects/plugins/jetpack/class.json-api-endpoints.php

```diff
@@ -2618,6 +2649,133 @@ public function get_amp_cache_origins( $siteurl ) {
 		);
 	}
 
+	/**
+	 * Register a REST route for this jsonAPI endpoint.
+	 *
+	 * @return void
+	 * @throws Exception The exception if something goes wrong.
+	 */
+	public function create_rest_route_for_endpoint() {
+		register_rest_route(
+			static::REST_NAMESPACE,
+			$this->build_rest_route(),
+			array(
+				'methods'             => $this->method,
+				'callback'            => array( $this, 'rest_callback' ),
+				'permission_callback' => array( $this, 'rest_permission_callback' ),
+			)
+		);
+	}
+
+	/**
+	 * Handle the rest call.
+	 *
+	 * @param WP_REST_Request $request The request object.
+	 *
+	 * @return mixed|WP_Error
+	 */
+	public function rest_callback( WP_REST_Request $request ) {
+		$blog_id = Jetpack_Options::get_option( 'id' );
+
+		$this->api->initialize();
+		$this->api->endpoint = $this;
+
+		$response = call_user_func_array(
+			array( $this, 'callback' ),
+			array_values( array( $this->path, $blog_id ) + $request->get_url_params() )
+		);
+
+		$token_data = ( new Manager() )->verify_xml_rpc_signature();
+
+		if ( ! $token_data || empty( $token_data['token_key'] ) || ! array_key_exists( 'user_id', $token_data ) ) {
+			return new WP_Error( 'response_signature_error' );
+		}
+
+		$token = ( new Tokens() )->get_access_token( $token_data['user_id'], $token_data['token_key'] );
+		if ( is_wp_error( $token ) ) {
+			return $token;
+		}
+		if ( ! $token ) {
+			return new WP_Error( 'response_signature_error' );
+		}
+
+		$response = wp_json_encode( $response );
+		$nonce    = wp_generate_password( 10, false );
+		$hmac     = hash_hmac( 'sha1', $nonce . $response, $token->secret );
+
+		return array(
+			$response,
+			(string) $nonce,
+			(string) $hmac,
+		);
+	}
+
+	/**
+	 * The REST endpoint should only be available for requests signed with a valid blog or user token.
+	 * Declaring it "final" so individual endpoints couldn't remove this requirement.
+	 *
+	 * @return true|WP_Error
+	 */
+	final public function rest_permission_callback() {
+		$manager = new Manager( 'jetpack' );
+		if ( ! $manager->is_connected() ) {
+			return new WP_Error( 'site_not_connected' );
+		}
+
+		$user_id          = get_current_user_id();
+		$allow_blog_token = $this->allow_fallback_to_jetpack_blog_token || $this->allow_jetpack_site_auth;
+
+		if ( ( $allow_blog_token && Rest_Authentication::is_signed_with_blog_token() ) || ( $user_id && Rest_Authentication::is_signed_with_user_token() ) ) {
```

Comment:
> We don't need to check `allow_fallback_to_jetpack_blog_token`

Good point, removed in https://github.com/Automattic/jetpack/pull/39432/commits/e97dd90f7a1b60715dfdb530ed220556ed039fa6

> Also do we really need to check Rest_Authentication::is_signed_with_user_token()? Wouldn't it be enough to check if $user_id is not empty assuming the request is signed with a user token?

We don't really want to open those endpoints for requests coming from logged in users.
So we check if this actually is user token based auth.

Reply:
> Good point, removed in https://github.com/Automattic/jetpack/commit/e97dd90f7a1b60715dfdb530ed220556ed039fa6

I believe we can use `$this->accepts_site_based_authentication` instead of `$this->allow_jetpack_site_auth && Rest_Authentication::is_signed_with_blog_token() `

Reply:
That's a good idea, but I'm not quite sure we should do that:
1. `token_details` [will have `blog_id` defined](https://github.com/Automattic/jetpack/blob/78d40351ea956cf913ab079e80b460798dc331d4/projects/plugins/jetpack/class.json-api.php#L346) for both user and blog token, so `is_jetpack_authorized_for_site()` will essentially use `get_current_user_id()` to determine the type of token (blog/user).
`is_signed_with_blog_token()` takes that information [directly from token](https://github.com/Automattic/jetpack/blob/89f781260f91d4acff0467d33b4a8242a7f2de37/projects/packages/connection/src/class-manager.php#L448), so I find that more straightforward and reliable.

2. We initialize JSON API [in the callback](https://github.com/Automattic/jetpack/blob/79fdce830c39d958b012d4aec0a1d0d8b571d94e/projects/plugins/jetpack/class.json-api-endpoints.php#L2680). That means that `token_details` is empty during the REST endpoint permission callback, and I'd rather avoid that.

3. I also don't feel confident about relying on `public_token` property because of this line: https://github.com/Automattic/jetpack/blob/8abca0d4b2270b06f5aa124368f26416f0bc3aa8/projects/plugins/jetpack/class-jetpack-xmlrpc-methods.php#L181
`$user_details` here comes from WPCOM directly, and trusted only because the request is signed. It's not something that comes from the token.
While it directly doesn't affect the blog token check in question, it's still something I'd rather stay away from.

Reply:
In this case, we should probably update  `accepts_site_based_authentication` then and re-use it everywhere, no?

Reply:
It's a widely used method already, so editing it may have unintended side effects in such a sensitive component as JSON API.
So I wouldn't want to do that.

Reply:
My point is, we're already doing some major changes here, so I'd prefer to minimize potential breaking points where we can.
So I'd rather avoid excessive refactoring, and revisit it separately later on.